src/server.rs

```diff
@@ -817,82 +850,175 @@ where
-                    match metadata(&archive)
+
+
+        let env = env.iter().cloned().collect::<Vec<(OsString,OsString)>>();
+
+        let me = self.clone();
+        let me1 = self.clone();
+        // TODO make the cache hold Box<dyn CompilerProxy>>
+        // TODO make everything else dependent on that result
+        // TODO and use it's methods as needed to resolve
+        // TODO to the actual compiler path
+
+        // lookup if compiler proxy exists for the current compiler path
+        let compiler_proxies_borrow = self.compiler_proxies.borrow();
+
+        let path2 = path.clone();
+        let path1 = path.clone();
+
+        let resolve_w_proxy = if let Some((compiler_proxy, filetime)) = compiler_proxies_borrow.get(&path) {
+            let fut = compiler_proxy.resolve_proxied_executable(
+                self.creator.clone(),
+                cwd.clone(),
+                env.as_slice(),
+            );
+            Box::new(fut)
+        } else {
+            f_ok(None)
+        };
+
+        drop(compiler_proxies_borrow);
+
+        // use the supplied compiler path as fallback, lookup it's modification time too
+        //
+        // TODO figure out a better way so in the transition phase with warmed caches this does
+        let w_fallback = resolve_w_proxy
+            .then(move |res: Result<Option<(PathBuf, FileTime)>>| {
+                let opt = match res {
+                    Ok(Some(x)) => Some(x), // TODO resolve the path right away
+                    _ => {
+                        // fallback to using the path directly
+                        metadata(&path2)
                         .map(|attr| FileTime::from_last_modification_time(&attr))
-                    {
-                        Ok(mtime) => Some((archive, mtime)),
-                        _ => None,
+                        .ok()
+                        .map(move |filetime| {
+                            (path2.clone(),filetime)
+                        })
+                    }
+                };
+                f_ok(opt)
+            });
+
+        let lookup_compiler =
+            w_fallback.and_then(move |opt : Option<(PathBuf, FileTime)>| {
+                let (resolved_compiler_path, mtime)
+                    = opt.expect("Must contain sane data, otherwise mtime is not avail");
+
+
+                let dist_info = match me1.dist_client.get_client() {
+                    Ok(Some(ref client)) => {
+                        if let Some(archive) = client.get_custom_toolchain(&resolved_compiler_path) {
+                            match metadata(&archive)
+                                .map(|attr| FileTime::from_last_modification_time(&attr))
+                            {
+                                Ok(mtime) => Some((archive, mtime)),
+                                _ => None,
+                            }
+                        } else {
+                            None
+                        }
```

Comment:
Feels like this could be done more succinctly with some `.map()` calls.

Reply:
IIRC the `&archive` borrow by `metadata` does not allow returning it by value - avoids a clone :)

Reply:
Ah, I missed the `archive` use further down.  OK, I can see how that might get complicated.